<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sparsely-Connected Autoencoders (SCAs) for single cell RNAseq data mining supplementary data • SCAtutorial</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sparsely-Connected Autoencoders (SCAs) for single cell RNAseq data mining supplementary data">
<meta property="og:description" content="SCAtutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SCAtutorial</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SCAvignette.html">Sparsely-Connected Autoencoders (SCAs) for single cell RNAseq data mining supplementary data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SCAvignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Sparsely-Connected Autoencoders (SCAs) for single cell RNAseq data mining supplementary data</h1>
                        <h4 class="author">Alessandri, Cordero, Beccuti, Licheri, Arigoni, Olivero, Di Renzo, Sapino and Calogero</h4>
            
      
      
      <div class="hidden name"><code>SCAvignette.Rmd</code></div>

    </div>

    
    
<div id="section-1" class="section level1">
<h1 class="hasAnchor">
<a href="#section-1" class="anchor"></a><strong>Section 1</strong>
</h1>
<p>A data set, <strong>setA</strong> [<a href="https://academic.oup.com/gigascience/article/8/9/giz105/5565135"><em>Alessandri et al. 2019</em></a>], based on FACS purified cell types [<a href="https://www.nature.com/articles/ncomms14049"><em>Zheng et al. 2017</em></a>] was used to investigate the SCA behaviour.</p>
<div id="seta-100-cells-randomly-selected-for-each-cell-type" class="section level2">
<h2 class="hasAnchor">
<a href="#seta-100-cells-randomly-selected-for-each-cell-type" class="anchor"></a><strong>setA: 100 cells randomly selected for each cell type</strong>
</h2>
<ul>
<li><ol start="2" style="list-style-type: upper-alpha">
<li>B-cells (25K reads/cell),</li>
</ol></li>
<li><ol start="13" style="list-style-type: upper-alpha">
<li>Monocytes (100K reads/cell),</li>
</ol></li>
<li><p>(HSC) Stem cells (24.7K reads/cell),</p></li>
<li><p>(NK) Natural Killer cells (29K reads/cell),</p></li>
<li><ol start="14" style="list-style-type: upper-alpha">
<li>Naive T-cells (19K reads/cell)</li>
</ol></li>
</ul>
<p>SetA was previously used to estimate the strength of CSS metric [<a href="https://academic.oup.com/gigascience/article/8/9/giz105/5565135"><em>Alessandri et al. 2019</em></a>]. We clustered setA using all the clustering tools actually implemented in rCASC: tSne+k-mean [<a href="Approximated%20and%20User%20Steerable%20tSNE%20for%20Progressive%20Visual%20Analytics"><em>Pezzotti et al. 2017</em></a>], SIMLR [<a href="https://www.nature.com/articles/nmeth.4207"><em>Wang et al. 2017</em></a>], griph [<a href="https://www.nature.com/articles/s41586-019-1146-y"><em>Serra et al. 2019</em></a>], Seurat [<a href="https://www.nature.com/articles/nbt.4096"><em>Butler et al. 2018</em></a>], scanpy [<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0"><em>Wolf et al. 2018</em></a>] and SHARP [<a href="https://genome.cshlp.org/content/early/2020/01/28/gr.254557.119"><em>Wan et al. 2020</em></a>]. All tools but tSne+k-mean and scanpy provided very good and similar partition of the different cell types (see Figure below).</p>
<pre><code>## Warning: package 'knitr' was built under R version 3.6.2</code></pre>
<div class="figure" style="text-align: center">
<img src="img/5clusters.png" alt="SetA counts table analysis" width="100%"><p class="caption">
SetA counts table analysis
</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Raffaele A Calogero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
